flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/metadata/FlinkRelMdUniqueKeys.scala

```diff
@@ -135,8 +138,8 @@ class FlinkRelMdUniqueKeys private extends MetadataHandler[BuiltInMetadata.Uniqu
                 }
               case _ => // ignore
             }
-          //rename
-          case a: RexCall if a.getKind.equals(SqlKind.AS) &&
+          //rename or cast
+          case a: RexCall if (a.getKind.equals(SqlKind.AS) || isFidelityCast(a)) &&
```

Comment:
Do we have to add this?
There are some casts that break uniqueness. For example, if cast fails, it will return null.

Reply:
That's true. That's why we only pick the fidelity casts that just change the precision or nullabilities (not-null -> nullable).